### ISSUE TITLE
update the regex for better experience

### ISSUE THREAD

**Issue by Rafi-Riday** [UziTech/marked-katex-extension#166]
The output of `$300 $400` is not correct, its behaving `$300 $` as a katex, but you can update it changing the regex to behave like katex input only when there is no space attaching with `$` in `$...$`, also, are you using the `displayMode: true` for the displayMode? Its important to use the actual displayMode option cause otherwise some functions does not works.

check out the https://stackedit.io/app for feature ideas

**Comment by UziTech** [comment]
If you want to create a PR for the Regexp update that would be very helpful 😁👍

For `displayMode` you can set the options when you use the extension:

```
marked.use(markedKatex({
  displayMode: true
}));
```

**Comment by Rafi-Riday** [comment]
Hei, I think I can change the regex to work specifically for this. But another thing, is `displayMode` automatically applies to `$$...$$`?

**Comment by UziTech** [comment]
I see what you mean about displayMode. We could default it to true for block level.

**Comment by Rafi-Riday** [comment]
Hei, i also tried to render this inside `$$\n...\n$$` :
```latex
\begin{equation}
\begin{split}   a &=b+c\\
      &=e+f
\end{split}
\end{equation}
```
But this does not work here, is this for because we're not using `displayMode : true`?
This problem is solved when in you extensions `index.js`, in the `blockKatex` function, I extend the `options` params with `displayMode : true` like this :
```js
renderer(token) {
    return `${katex.renderToString(token.text, { ...options, displayMode: true })}`;
}
```
So I think this is why we need the `displayMode : true`, the official way to use katex.
I also changed the regexp, but I'm testing more, https://github.com/benweet/stackedit this vue.js project has implemented  some functionality for better user experience like avoiding the `$300 $400` problem etc. I'm also implementing some of those here, but you need to check them before merging

**Comment by UziTech** [comment]
I created PR #167 to fix this. See if that works for you.

**Comment by Rafi-Riday** [comment]
I am actually thinking of setting some rule, like `$...$` for inline mode, and `$$...$$` for display mode, so `$$$....$$$` will not be valid. This might fail some test in this project, but its more usual. also `$$...$$` and `$$\n...\n$$` both are for displayMode.

There is maybe some internal problem thats why 
```latex
$$
\colorbox{aqua}{$
\frac{b}{2}\sqrt{a^2-\frac{b^2}{4}}
a$}
$$
```
does not works.
Is there any problem if I make thease changes and then make PR?

**Comment by UziTech** [comment]
I think we should require escaping `$` in katex. Otherwise the regular expression might be difficult.

**Comment by UziTech** [comment]
I think it would be fine to make those changes. We will just need to release a major change if we limit to `$` for inline and `$$` for block.

**Comment by UziTech** [comment]
Just make sure you update the tests to pass and create enough tests to test each edgecase

**Comment by Rafi-Riday** [comment]
Yeah man i changed regex, `$` for inline works fine, but `$$` or more `$` will be considered as block, all the test are passed, but still doing some changes, and one more thing,
```latex
$$
\colorbox{aqua}{$
\frac{b}{2}\sqrt{a^2-\frac{b^2}{4}}
a$}
$$
```
This block contains `$` and this may break cause there are 2 katex match, I think I have to change it a little bit so that inline `$` are not being matched inside `$$` block. So the full text is passed to katex renderer, the regexp is a bit bigger but still easy to understand at this moment. just after this feature is added I'll create a PR after making sure all the tests are passsed.

Here is a demo where I am rendering `katex` without `marked` : https://katex-output.netlify.app/
The above code works, cause everything in the textarea goes, is supposed to be katex input. The regexp is maybe the main solve to fix thi katex error of this version of the extension.

**Comment by Rafi-Riday** [comment]
I think we should update extension to make `$...$` inline and `$$...$$` block level. the block level may contain multiline, but the inline level can not contain multiple lines. This will make users strictly use this way. if the inline katex is wrapped with `$$...$$` then it will behave like block.

**Comment by UziTech** [comment]
So `$$` would just mean `displayMode: true` even if it is inline? I think that sounds good. Ya I like making it strict. We can always relax it if a valid use case is found.

**Comment by Rafi-Riday** [comment]
Hei i made some improvements but no idea if i actually broke something. cause the test fails. But maybe they fail because `$\n...\n$` is now inline (we can remove this feature if we want). the `$\n...\n$` should not be a block, cause in katex, there is some use cases where we need to write `$...$` inside another katex. like this one :
```latex
$$
\colorbox{aqua}{$
\frac{b}{2}\sqrt{a^2-\frac{b^2}{4}}
$}
$$
```

In my version, the cod below works fine, but the previous code does not work
```latex
$$
\colorbox{aqua}{$\frac{b}{2}\sqrt{a^2-\frac{b^2}{4}}$}
$$
```
This is because of the strict type `$...$` and `$$...$$` ig.
But some test is failing because i changed most of the rules. But ig guess those rules are necessary in order to use katex, katex itself uses `$...$` for inline katex. like this code : `\text{$\pi$}`, if `\pi` is wrapped by double `$` then they wont render `\pi`, rather the output will be 2 empty katex output wrapping a text `\pi`. We must use `$...$` in this case.

Right now, my version of the code accepts `$\n...\n$` for inine katex, but we should remove `\n`. Also, For block level, 2 or more `$` works. Like `$$$...$$$`
Here is my preview : https://marked-katex-extension.netlify.app/
Please check it out if there is any edge cases, any bugs, so that I can fix them. Currently many test fails, but those are because i changed rules. I think we have to test with newer code. I hope this library will work more better after thease improvements

**Comment by UziTech** [comment]
Where is your code? I think I have #167 working so inline katex requires spaces before and after `$...$`. Inline `$$...$$` works with no newline chars and will be `displayMode: true`. Block level works with `$$\n...\n$$` and can contain newline chars.

I am working on adding more tests.

**Comment by Rafi-Riday** [comment]
#171

**Comment by Rafi-Riday** [comment]
I created #171 that works mostly, but see if it fails edge cases. Here I made support for `$\n...\n$` which is still inline, I think its horrible idea to be updated. But anything which is wrapped by `$$` or more `$` is block level katex. Even if its inline, it will render block level, which is better I think. Editors like stackedit.io uses this rules.
Now some other things like `$\text{$hi$ there $lol$}$` also works, but I dont know why, it didn't work in the currnet version of katex-extension. May e we have to add test for thease use cases, also we have to check everything from here : https://katex.org/docs/supported.html ,they are supported functions by katex, not everything supports by this extension currently, I think my code works just the extension is not rendering it properly.

We have to update this extension :
- add support for `\$` to escape and not to behave like a wrapper `$`,
- `$\n...\n$` has to be removed,
- block level wrapper will only contain `$$`, and if more `$` is wrapping, then the outter extra `$` will be matched as a normal `$` instead of a wrapper `$`,
- add support for multiline for the case where a wrapper is inside another wrapper, so the parent wrapper will be matched as a katex, and the child wrapper wont cause there is a parent. This will prevent collision of them.

I hope thease updates will make the extension a little bit strict but will add more support for katex
